src/impls.rs

```diff
@@ -113,200 +90,133 @@ impl_metadata_for_tuple!(A, B, C, D, E, F, G, H);
 impl_metadata_for_tuple!(A, B, C, D, E, F, G, H, I);
 impl_metadata_for_tuple!(A, B, C, D, E, F, G, H, I, J);
 
-impl<T> HasTypeId for Vec<T>
-where
-	T: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Vec", Namespace::prelude(), tuple_meta_type![T]).into()
-	}
-}
-
-impl<T> HasTypeDef for Vec<T>
+impl<T> HasType for Vec<T>
 where
 	T: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[T]>())]).into()
+	fn get_type() -> Type {
+		product_type(
+			"Vec",
+			Namespace::prelude(),
+			tuple_meta_type![T],
+			TypeProductStruct::new(vec![NamedField::new("elems", MetaType::new::<[T]>())]),
+		)
 	}
 }
 
-impl<T> HasTypeId for Option<T>
+impl<T> HasType for Option<T>
 where
 	T: Metadata + 'static,
 {
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Option", Namespace::prelude(), tuple_meta_type![T]).into()
+	fn get_type() -> Type {
+		sum_type(
+			"Option",
+			Namespace::prelude(),
+			tuple_meta_type![T],
+			TypeSumEnum::new(vec![
+				EnumVariantUnit::new("None").into(),
+				EnumVariantTupleStruct::new("Some", vec![UnnamedField::of::<T>()]).into(),
+			]),
+		)
 	}
 }
 
-impl<T> HasTypeDef for Option<T>
-where
-	T: Metadata + 'static,
-{
-	fn type_def() -> TypeDef {
-		TypeDefEnum::new(vec![
-			EnumVariantUnit::new("None").into(),
-			EnumVariantTupleStruct::new("Some", vec![UnnamedField::of::<T>()]).into(),
-		])
-		.into()
-	}
-}
-
-impl<T, E> HasTypeId for Result<T, E>
-where
-	T: Metadata + 'static,
-	E: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Result", Namespace::prelude(), tuple_meta_type!(T, E)).into()
-	}
-}
-
-impl<T, E> HasTypeDef for Result<T, E>
+impl<T, E> HasType for Result<T, E>
 where
 	T: Metadata + 'static,
 	E: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefEnum::new(vec![
-			EnumVariantTupleStruct::new("Ok", vec![UnnamedField::of::<T>()]).into(),
-			EnumVariantTupleStruct::new("Err", vec![UnnamedField::of::<E>()]).into(),
-		])
-		.into()
-	}
-}
-
-impl<K, V> HasTypeId for BTreeMap<K, V>
-where
-	K: Metadata + 'static,
-	V: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("BTreeMap", Namespace::prelude(), tuple_meta_type!(K, V)).into()
+	fn get_type() -> Type {
+		sum_type(
+			"Result",
+			Namespace::prelude(),
+			tuple_meta_type!(T, E),
+			TypeSumEnum::new(vec![
+				EnumVariantTupleStruct::new("Ok", vec![UnnamedField::of::<T>()]).into(),
+				EnumVariantTupleStruct::new("Err", vec![UnnamedField::of::<E>()]).into(),
+			]),
+		)
 	}
 }
 
-impl<K, V> HasTypeDef for BTreeMap<K, V>
+impl<K, V> HasType for BTreeMap<K, V>
 where
 	K: Metadata + 'static,
 	V: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[(K, V)]>())]).into()
-	}
-}
-
-impl<T> HasTypeId for Box<T>
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn get_type() -> Type {
```

Comment:
rename to just `type` please or we rename the whole trait to `MetaType` or `TypeInfo` or so and rename the method to `meta_type` or `type_info` respectively. `HasType` is actually a bad name but I used it since `Type` was too ubiquitous.

Reply:
Gone for `TypeInfo`